schema.development.kf

```diff
@@ -637,69 +611,136 @@ action share_credential_through_dag (
     );
 }
 
-action share_credential_by_write_grant (
-    $id,
-    $user_id,
+action create_credentials_by_dwg(
+    $issuer_auth_public_key,
     $encryptor_public_key,
-    $content,
-    $content_hash,
-    $public_notes,
-    $public_notes_signature,
-    $broader_signature,
     $original_credential_id,
-    $grantee_wallet_identifier,
-    $locked_until,
-    $issuer_auth_public_key
-) public {
-    SELECT CASE WHEN $locked_until < 0 THEN
-        error('locked_until must be positive integer timestamp or zero')
-    END;
-
-    SELECT get_write_grant_id($user_id); // throw an error if no write_grant
+    $original_content,
+    $original_public_notes,
+    $original_public_notes_signature,
+    $original_broader_signature,
+    $copy_credential_id,
+    $copy_content,
+    $copy_public_notes_signature, // Signature of an empty string?! This is weird.
+    $copy_broader_signature,
+    $content_hash, // For access grant
+    $dwg_owner,
+    $dwg_grantee,
+    $dwg_id,
+    $dwg_access_grant_timelock,
+    $dwg_not_before,
+    $dwg_not_after,
+    $dwg_signature) public {
 
-    SELECT CASE WHEN NOT credential_belongs_to_user($original_credential_id, $user_id) THEN
-        error('the original credential does not belong to the user')
+    SELECT CASE
+        WHEN $issuer_auth_public_key != $dwg_grantee
+        THEN ERROR('credentials issuer must be a grantee of delegated write grant (issuer_auth_public_key = dwg_grantee)')
     END;
 
-    SELECT CASE WHEN get_write_grant_id($user_id)::text != get_credential_inserter($original_credential_id) THEN
-        error('you can share only original credentials you created')
+    SELECT CASE
+        WHEN NOT EXISTS (SELECT 1 FROM wallets WHERE (wallet_type = 'EVM' AND address=$dwg_owner COLLATE NOCASE)
+            OR (wallet_type = 'NEAR' AND public_key = $dwg_owner))
+        THEN ERROR('dwg_owner not found')
     END;
 
-    SELECT CASE WHEN (SELECT count(1) FROM access_grants WHERE id = uuid_generate_v5('31276fd4-105f-4ff7-9f64-644942c14b79'::uuid, format('%s-%s-%s', $grantee_wallet_identifier::text, $id::text, $locked_until::text))) > 0 THEN
-        error('a grant with the same grantee, original_credential_id, and locked_until already exists')
+    $ag_timelock = idos.parse_date($dwg_access_grant_timelock); // Will fail if not in the RFC3339 format
+    $times_validation = idos.validate_not_usable_times($dwg_not_before, $dwg_not_after); // Check the format and precedence
+    SELECT CASE WHEN $times_validation != 1 THEN ERROR('dwg_not_before must be before dwg_not_after') END;
```

Comment:
Since we're parsing the strings anyway, why not compare the results instead of calling into an extension function?

Reply:
`validate_not_usable_times` does validate `dwg_not_before` and `dwg_not_after` formats and then compare them.
We can do comparison in the schema, but we anyway need to call the extension to validate dates formats.
What approach do you prefer?

Reply:
Hmm, yeah, I get your point. Lemme experiment with `parse_unix_timestamp`.

Reply:
Hmm, things don't really explode like I was expecting them to:
``` bash
$ kwil-cli database query -n idos "select parse_unix_timestamp('2022-01-01S01:01:02X', 'YYYY-MM-DD"'"'"T"'"'"HH24:MI:SS"'"'"Z"'"'"') as derp"
|       derp        |
+-------------------+
| 1640998862.000000 |
```

I don't think it's worth our time to iterate this further.